src/bin/pgcopydb/pgsql_timeline.c

```diff
@@ -0,0 +1,513 @@
+		return;
+	}
+
+	if (PQntuples(result) == 0)
+	{
+		log_sql("parseIdentifySystem: query returned no rows");
+		context->parsedOk = false;
+		return;
+	}
+	if (PQntuples(result) != 1)
+	{
+		log_error("Query returned %d rows, expected 1", PQntuples(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	/* systemid (text) */
+	char *value = PQgetvalue(result, 0, 0);
+	if (!stringToUInt64(value, &(context->system->identifier)))
+	{
+		log_error("Failed to parse system_identifier \"%s\"", value);
+		context->parsedOk = false;
+		return;
+	}
+
+	/* timeline (int4) */
+	value = PQgetvalue(result, 0, 1);
+	if (!stringToUInt32(value, &(context->system->timeline)))
+	{
+		log_error("Failed to parse timeline \"%s\"", value);
+		context->parsedOk = false;
+		return;
+	}
+
+	/* xlogpos (text) */
+	value = PQgetvalue(result, 0, 2);
+	strlcpy(context->system->xlogpos, value, PG_LSN_MAXLENGTH);
+
+	/* dbname (text) Database connected to or null */
+	if (!PQgetisnull(result, 0, 3))
+	{
+		value = PQgetvalue(result, 0, 3);
+		strlcpy(context->system->dbname, value, NAMEDATALEN);
+	}
+
+	context->parsedOk = true;
+}
+
+
+/*
+ * parseTimelineHistoryResult parses the result of the TIMELINE_HISTORY
+ * replication command. The content is written to disk, and the filename
+ * is stored in the TimelineHistoryResult structure.
+ */
+static void
+parseTimelineHistoryResult(TimelineHistoryResult *context, PGresult *result,
+						   char *cdcPathDir)
+{
+	if (PQnfields(result) != 2)
+	{
+		log_error("Query returned %d columns, expected 2", PQnfields(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	if (PQntuples(result) == 0)
+	{
+		log_sql("parseTimelineHistoryResult: query returned no rows");
+		context->parsedOk = false;
+		return;
+	}
+
+	if (PQntuples(result) != 1)
+	{
+		log_error("Query returned %d rows, expected 1", PQntuples(result));
+		context->parsedOk = false;
+		return;
+	}
+
+	/* filename (text) */
+	char *value = PQgetvalue(result, 0, 0);
+	strlcpy(context->filename, value, sizeof(context->filename));
+
+	/*
+	 * content (bytea)
+	 *
+	 * We do not want to store this value in memory. Instead we write it to disk
+	 * as it is.
+	 */
+	value = PQgetvalue(result, 0, 1);
+	if (!writeTimelineHistoryFile(context->filename, value, cdcPathDir))
+	{
+		log_error("Failed to write timeline history file \"%s\"", context->filename);
+		context->parsedOk = false;
+		return;
+	}
+
+	context->parsedOk = true;
+}
+
+
+/* timeline_iter_history is used to iterate over the content of a timeline history file. */
+bool
+timeline_iter_history(char *filename,
+					  ParseTimelineHistoryContext *context,
+					  TimelineHistoryFun *callback)
+{
+	TimelineHistoryIterator *iter =
+		(TimelineHistoryIterator *) malloc(sizeof(TimelineHistoryIterator));
+
+	if (iter == NULL)
+	{
+		log_error("Failed to allocate memory for timeline history iterator");
+		return false;
+	}
+
+	iter->filename = filename;
+	iter->currentTimeline = context->currentTimeline;
```

Comment:
The iterator should never look into the context.

Reply:
This is a great point. Applying a fix for this allowed me to make the following improvements:
- I passed the value as an argument, I had a slimmer context.
- Now that the context had a single attribute, I deleted that completely.
- A function with a bad name `timeline_history_add_hook` that was a wrapper for `catalog_add_timeline_history` was no longer needed as I did not need to have the logic to access a subset of values in a context.

When the number of parameters get large, I start to think of context structs as a means to group them. This was a wrong way to think, and I see that now.